sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+       * <a href="https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-config.html">Configuring the KPL</a>
+       */
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        checkForFailures();
+
+        // Need to avoid keeping too many futures in producer's map to prevent OOM.
+        // In usual case, it should exit immediately.
+        flush(MAX_NUM_RECORDS);
+
+        ByteBuffer data = ByteBuffer.wrap(c.element());
+        String partitionKey = spec.getPartitionKey();
+        String explicitHashKey = null;
+
+        // Use custom partitioner
+        if (partitioner != null) {
+          partitionKey = partitioner.getPartitionKey(c.element());
+          explicitHashKey = partitioner.getExplicitHashKey(c.element());
         }
-        return clientBuilder.build();
+
+        ListenableFuture<UserRecordResult> f =
+            producer.addUserRecord(spec.getStreamName(), partitionKey, explicitHashKey, data);
+        Futures.addCallback(f, new UserRecordResultFutureCallback());
       }
 
-      @Override
-      public AmazonCloudWatch getCloudWatchClient() {
-        AmazonCloudWatchClientBuilder clientBuilder =
-            AmazonCloudWatchClientBuilder.standard().withCredentials(getCredentialsProvider());
-        if (serviceEndpoint == null) {
-          clientBuilder.withRegion(region);
-        } else {
-          clientBuilder.withEndpointConfiguration(
-              new AwsClientBuilder.EndpointConfiguration(serviceEndpoint, region.getName()));
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        checkForFailures();
+
+        // Flush all outstanding records, blocking call
+        flushAll();
+      }
+
+      @Teardown
+      public void tearDown() throws Exception {
+        if (producer != null) {
+          producer.destroy();
+          producer = null;
+        }
+      }
+
+      /**
+       * Flush outstanding records until the total number will be less than required or
+       * number of retries will be exhausted.
+       */
+      private void flush(int numMax) throws InterruptedException, IOException {
+        int retries = spec.getRetries();
+        int numOutstandingRecords = producer.getOutstandingRecordsCount();
+
+        while (numOutstandingRecords > numMax && retries-- > 0) {
+          producer.flush();
+          // wait until outstanding records will be flushed
+          Thread.sleep(1000);
+          numOutstandingRecords = producer.getOutstandingRecordsCount();
+        }
+
+        if (numOutstandingRecords > numMax) {
+          String message =
+              String.format(
+                  "Number of outstanding records [%d] are greater than required [%d].",
+                  numOutstandingRecords,
+                  numMax);
+          LOG.error(message);
+          throw new IOException(message);
```

Comment:
I'm not sure we should fail here. We might be able to proceed without losing data; if an OOM happens, there'll be plenty of logging to explain why.

Reply:
As I mentioned before, I use this method also to flush all bundle records in `FinishBundle`, so it makes sense to fail. Probably, we can change this logic and fail only in `FinishBundle` (or even in `TearDown`). 
What do you think?

Reply:
Oh right. It would make sense to fail only in `FinishBundle`, but probably not worth the trouble.

Reply:
Yes, let's keep it "as it is" for the moment.